### PR TITLE
Fix for bug showing incorrect awareness attributes count in AwarenessAllocationDecider

### PR DESCRIPTION
**Pull request by anshu1106**
…Decider

This commit fixes allocation explanation reason displaying incorrect number of awareness attributes in some cases when forced zone awareness is enabled.

Signed-off-by: Anshu Agarwal <anshukag@amazon.com>

### Description
This commit fixes bug in AwarenessAllocationDecider showing incorrect number of awareness attribute in allocation explanation reason. More details on the bug and how to reproduce it can be found in the issue.

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/3413
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/test/java/org/opensearch/cluster/routing/allocation/AwarenessAllocationTests.java [hunk 1]
```diff
@@ -35,23 +35,23 @@
...
-import org.opensearch.cluster.routing.ShardRouting;
-import org.opensearch.cluster.routing.ShardRoutingState;
+import org.opensearch.cluster.routing.*;
 import org.opensearch.cluster.routing.allocation.command.AllocationCommands;
 import org.opensearch.cluster.routing.allocation.command.CancelAllocationCommand;
 import org.opensearch.cluster.routing.allocation.command.MoveAllocationCommand;
-import org.opensearch.cluster.routing.allocation.decider.AwarenessAllocationDecider;
-import org.opensearch.cluster.routing.allocation.decider.ClusterRebalanceAllocationDecider;
-import org.opensearch.cluster.routing.allocation.decider.ThrottlingAllocationDecider;
+import org.opensearch.cluster.routing.allocation.decider.*;
```
Comment (reta) [inline server/src/test/java/org/opensearch/cluster/routing/allocation/AwarenessAllocationTests.java:48]: Please revert the change to use `*`, the concrete types are preferred.

**Review by nknize — DISMISSED** [review]
Rebase from main.

File: server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java [hunk 2]
```diff
@@ -209,7 +209,7 @@ private Decision underCapacity(ShardRouting shardRouting, RoutingNode node, Rout
             List<String> fullValues = forcedAwarenessAttributes.get(awarenessAttribute);
             if (fullValues != null) {
                 for (String fullValue : fullValues) {
-                    if (shardPerAttribute.containsKey(fullValue) == false) {
+                    if (nodesPerAttribute.containsKey(fullValue) == false) {
```
[the remark this answers is not in the document]
Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java:212]: Can we simplify the logic to reflect 
```
ceil(count(union(discovered_zone, force_zone))/total_shard_copies)
```

File: server/src/test/java/org/opensearch/cluster/routing/allocation/AwarenessAllocationTests.java (hunk 1 above)
Comment (Bukhtawar) [reply server/src/test/java/org/opensearch/cluster/routing/allocation/AwarenessAllocationTests.java:48]: +1

File: server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java:212]: > Could you please explain this change? AFAIK, the allocation is about shards, not nodes, as such usage of `shardPerAttribute` is about to be right (to me).

Yeah, thats true. But this implementation is showing incorrect attribute count in some cases when forced awareness is enabled for the domain which I mentioned in the issue. I have created a unit test which fails as well with incorrect awareness attribute count. 
           ```
 int numberOfAttributes = nodesPerAttribute.size();
            List<String> fullValues = forcedAwarenessAttributes.get(awarenessAttribute);
            if (fullValues != null) {
                for (String fullValue : fullValues) {
                    if (shardsPerAttribute.containsKey(fullValue) == false) {
                        numberOfAttributes++;
                    }
                }
            }
```

In code snippet, we are first calculating number of awareness attributes using discovered zones. Assume the domain has 3 zones but one zone (zone-c) is completely down, then nodesPerAttribute.size() returns 2. 
Assume forced awareness value is set to  "zone-a, zone-b, zone-c". So this returns three. ie fullValues =3. 

 Now assume a node came up in zone-c which had outage earlier. But no shards has been assigned to that node till now. So in this case shardsPerAttribute.size()=2 since there are no shards assigned to any node in zone-c but nodesPerAttribute.size()=3. Now when we go inside inner for loop, shardsPerAttribute won't have key for zone-c and it increases numberOfAttribute count to 4 which is incorrect.(ie zone-c has been counted twice)
So basically numAttributes = distinct(union(discovered_zone, forced_zone))

[the remark this answers is not in the document]
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java:212]: > Could you please explain this change? AFAIK, the allocation is about shards, not nodes, as such usage of `shardPerAttribute` is about to be right (to me).

@reta  Yeah, thats true. But this implementation is showing incorrect attribute count in some cases when forced awareness is enabled for the domain which I mentioned in the issue. I have created a unit test which fails as well with incorrect awareness attribute count.

`

            int numberOfAttributes = nodesPerAttribute.size();
            List<String> fullValues = forcedAwarenessAttributes.get(awarenessAttribute);
            if (fullValues != null) {
                for (String fullValue : fullValues) {
                    if (shardPerAttribute.containsKey(fullValue) == false) {
                        numberOfAttributes++;
                    }
                }
            }
`

In code snippet, we are first calculating number of awareness attributes using discovered zones. Assume the domain has 3 zones but one zone (zone-c) is completely down, then nodesPerAttribute.size() returns 2. 
Assume forced awareness value is set to  "zone-a, zone-b, zone-c". So this returns three. ie fullValues =3. 

Now assume a node came up in zone-c which had outage earlier. But no shards has been assigned to that node till now. So in this case shardPerAttribute.size()=2 since there are no shards assigned to any node in zone-c but nodesPerAttribute.size()=3. Now when we go inside inner for loop, shardPerAttribute won't have key for zone-c and it increases numberOfAttribute count to 4 which is incorrect.(ie zone-c has been counted twice)
So basically numAttributes = distinct(union(discovered_zone, forced_zone))

[the remark this answers is not in the document]
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java:212]: > Can we simplify the logic to reflect
> 
> ```
> ceil(count(union(discovered_zone, force_zone))/total_shard_copies)
> ```

Sure will try to make the logic simpler.

[the remark this answers is not in the document]
Comment (reta) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java:212]: I think it makes sense now, thank @anshu1106 !

File: server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java [hunk 3]
```diff
@@ -207,12 +210,14 @@ private Decision underCapacity(ShardRouting shardRouting, RoutingNode node, Rout
...
             int numberOfAttributes = nodesPerAttribute.size();
             List<String> fullValues = forcedAwarenessAttributes.get(awarenessAttribute);
+            Set<String> attributesSet = new HashSet<>();
             if (fullValues != null) {
-                for (String fullValue : fullValues) {
-                    if (shardPerAttribute.containsKey(fullValue) == false) {
-                        numberOfAttributes++;
-                    }
+                // If forced awareness is enabled, numberOfAttributes = count(distinct((union(discovered_attributes, forced_attributes)))
+                attributesSet.addAll(forcedAwarenessAttributes.get(awarenessAttribute));
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java:216]:
```suggestion
                attributesSet.addAll(fullValues);
```

File: server/src/test/java/org/opensearch/cluster/routing/allocation/AwarenessAllocationTests.java [hunk 4]
```diff
@@ -971,4 +980,87 @@ public void testMultipleAwarenessAttributes() {
...
+        final AwarenessAllocationDecider decider = new AwarenessAllocationDecider(settings, EMPTY_CLUSTER_SETTINGS);
+
+        final RoutingAllocation allocation = new RoutingAllocation(
+            new AllocationDeciders(Collections.singleton(decider)),
+            clusterState.getRoutingNodes(),
+            clusterState,
+            null,
+            null,
+            0L
+        );
```
Comment (Bukhtawar) [inline server/src/test/java/org/opensearch/cluster/routing/allocation/AwarenessAllocationTests.java:1044]: Why do we need to have only a single decide here?

File: server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java (hunk 3 above)
Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java:216]: Lets modify this accordingly

File: server/src/test/java/org/opensearch/cluster/routing/allocation/AwarenessAllocationTests.java (hunk 4 above)
Comment (anshu1106) [reply server/src/test/java/org/opensearch/cluster/routing/allocation/AwarenessAllocationTests.java:1044]: This test is to check allocation decision by AwarenessAllocationDecider Line no -1053. Wanted the test to be more predictable, thats why didn't introduce all the deciders.

**Review by Bukhtawar — APPROVED** [review]
